turbopack/crates/turbo-tasks-macros/src/func.rs

```diff
@@ -584,7 +627,21 @@ impl TurboFn<'_> {
                     )
                 }
             }
+        };
+        if self.operation {
+            block = parse_quote! {
+                {
+                    let vc_output = #block;
+                    // SAFETY: The turbo-tasks manager will not create a local task for a function
+                    // where all task inputs are "resolved" (where "resolved" in this case includes
+                    // `OperationVc`). This is checked with a debug_assert, but not in release mode.
```

Comment:
This shouldn't use `unsafe`.

Reply:
This is reserving the right to remove a few safety checks in the future on release builds for performance reasons:

- You could stuff a local `Vc` into a `VcOperation` type on a release build, because the constructor's check is debug-only.
- When reading a local `Vc`, we check the execution id (this bloats the size of the `Vc` type) to ensure we're reading from the correct task execution array. Local cell storage uses a flat vec (containing type ids, but not indexed by type id), which could mean that if a task leaks and we remove the execution id in the future, we might read a cell of the wrong type from a different task execution and try to downcast it incorrectly.
- `ReadVcFuture` currently uses `TypedCellContent` which performs a checked downcast. This should never reasonably fail in the context of `ReadVcFuture`, so it could be an unsafe unchecked downcast in the future.
- We could optimize `connect()` in the future by using a match block that assumes the representation of the `RawVc` by calling [`unreachable_unchecked`](https://doc.rust-lang.org/std/hint/fn.unreachable_unchecked.html) on non-`TaskOutput` branches.

We always expect local cells to be consistent with the `Vc`'s type `T`, so there's some possibility that could later lead to a wrong transmute when we read the cell from the wrong `local_cells` vec.

However, you're right that this isn't *currently* unsafe, so I'll just remove it and mark the method as `deprecated` to further discourage use. Since this should only be used inside of macros, it shouldn't be hard to make it unsafe in the future.